contracts/introspection/ERC165Checker.sol

```diff
@@ -0,0 +1,111 @@
+   *      with the `supportsERC165` method in this library. 
+   *      Interface identification is specified in ERC-165.
+   */
+  function supportsERC165Interface(address _address, bytes4 _interfaceId)
+    internal
+    view
+    returns (bool)
+  {
+    // success determines whether the staticcall succeeded and result determines
+    // whether the contract at _address indicates support of _interfaceId
+    (bool success, bool result) = noThrowCall(_address, _interfaceId);
+
+    return (success && result);
+  }
+
+  /**
+   * @notice Query if a contract supports ERC165
+   * @param _address The address of the contract to query for support of ERC165
+   * @return true if the contract at _address implements ERC165
+   */
+  function supportsERC165(address _address)
+    internal 
+    view 
+    returns (bool)
+  {
+    // Any contract that implements ERC165 must explicitly indicate support of
+    // InterfaceId_ERC165 and explicitly indicate non-support of InterfaceId_Invalid
+    return supportsERC165Interface(_address, InterfaceId_ERC165) &&
+      !supportsERC165Interface(_address, InterfaceId_Invalid);
+  }
+
+  /**
+   * @notice Query if a contract implements an interface, also checks support of ERC165
+   * @param _address The address of the contract to query for support of an interface
+   * @param _interfaceId The interface identifier, as specified in ERC-165
+   * @return true if the contract at _address indicates support of the interface with
+   * identifier _interfaceId, false otherwise
+   * @dev Interface identification is specified in ERC-165.
+   */
+  function supportsInterface(address _address, bytes4 _interfaceId)
+    internal
+    view
+    returns (bool)
+  {
+    // query support of both ERC165 as per the spec and support of _interfaceId
+    return supportsERC165(_address) &&
+      supportsERC165Interface(_address, _interfaceId);
+  }
+
+  /**
+   * @notice Calls the function with selector 0x01ffc9a7 (ERC165) and suppresses throw
+   * @param _address The address of the contract to query for support of an interface
+   * @param _interfaceId The interface identifier, as specified in ERC-165
+   * @return success true if the STATICCALL succeeded, false otherwise
+   * @return result true if the STATICCALL succeeded and the contract at _address
+   * indicates support of the interface with identifier _interfaceId, false otherwise
+   */
+  function noThrowCall (
+    address _address,
+    bytes4 _interfaceId
+  ) 
+    internal
+    view
+    returns (bool success, bool result)
+  {
+    bytes4 erc165ID = InterfaceId_ERC165;
+
+    // solium-disable-next-line security/no-inline-assembly
+    assembly {
+        let x := mload(0x40)               // Find empty storage location using "free memory pointer"
+        mstore(x, erc165ID)                // Place signature at begining of empty storage
+        mstore(add(x, 0x04), _interfaceId) // Place first argument directly next to signature
```

Comment:
I think this can be done with `abi.encodeWithSelector`, and thus remove at least part of the assembly code. Also I'm not a fan of the `x` variable name. :sweat_smile:

Reply:
If i understand correctly, you would prefer something like:

```
  function noThrowCall (
    address _address,
    bytes4 _interfaceId
  ) 
    internal
    view
    returns (bool success, bool result)
  {
    // encode erc165 selector and first argument (_interfaceId)
    bytes memory encodedParams = abi.encodeWithSelector(InterfaceId_ERC165, _interfaceId);

    // solium-disable-next-line security/no-inline-assembly
    assembly {
        let freeMemoryPtr := mload(0x40)     // Find empty storage location using "free memory pointer"
        mstore(freeMemoryPtr, encodedParams) // Place encoded args at beginning of empty storage

        success := staticcall(
                  30000,         // 30k gas
                  _address,      // To addr
                  freeMemoryPtr, // Inputs are stored at free memory pointer
                  0x20,          // Inputs are 32 bytes long
                  freeMemoryPtr, // Store output over input (saves space)
                  0x20)          // Outputs are 32 bytes long

        result := mload(freeMemoryPtr) // Load the result
    }
  }
```

I just tested and this does not work (tests fail). The [documentation](http://solidity.readthedocs.io/en/latest/units-and-global-variables.html#abi-encoding-functions) implies that this should be equivalent to the previous code, but I'm probably misunderstanding something. Can you spot the bug?

I agree that it looks cleaner and am happy to push a change to this PR once I can fix this bug.

@nventuro @shrugs thoughts on this change?

Reply:
The `encode` call gives you back a memory pointer to a dynamic byte array. It's already in memory so you don't need to manually `mstore` it. However, the dynamic byte array has a 32 byte length before the contents, so you should use `add(encodedParams, 4)` where `freeMemoryPtr` is being used now. I'm not a fan of storing the output over the input, but I suppose it works.

On second thought, I have a few more doubts about it. Are we sure that the result of `abi.encodeWithSelector` is 32 bytes long?

Reply:
Tried what you suggested and it didn't work either.

Spent some time trying to use velma debugger to inspect what exactly abi.encodeWithSelector is giving (as the documentation is kind of sparse on this), but can't seem to get it to work with solidity 0.4.24. Will try with truffle debug and/or hevm tomorrow.

Reply:
@ldub I was able to get `abi.encodeWithSelector` working on 0.4.24 as follows:

```solidity
bytes memory encodedParams = abi.encodeWithSelector(InterfaceId_ERC165, _interfaceId);

// solium-disable-next-line security/no-inline-assembly
assembly {
    let encodedParams_data := add(0x20, encodedParams)
    let encodedParams_size := mload(encodedParams)
    
    let output := mload(0x40)               // Find empty storage location using "free memory pointer"

    success := staticcall(
              30000,     // 30k gas
              _address,  // To addr
              encodedParams_data,
              encodedParams_size,
              output,
              0x20)      // Outputs are 32 bytes long

    result := mload(output)   // Load the result
}
```